### DISCUSSION TITLE
Clarification on using new Client()

### DISCUSSION THREAD

**Discussion by NN-Binary** [nodejs/undici#1040] [Q&A]
I would like to control my own 'keep-alive' as in re-using connections whenever I can and be in control of it, currently using the Undici.dispatch way, I would like to know how should I start? I'm thinking of re-using the same undici.Client for each hostname and put them in a Map and just retrieve them.

MAP[Hostname] = new undici.Client('https://www.google.com'); 

My main question is: If I'm doing so, will it be enough to guarantee that I'm re-using the same TCP connection? I would like to avoid the TLS handshake all the time but being able to just request anything (via dispatch Undici method) in the coming seconds without having to reconnect.

Thank you, still learning and I'm having difficulties

**Comment by paganwinter** [comment]
I'm interested in knowing about this as well!

**Reply by mcollina** [reply]
The usage is correct. Client guarantees that only one socket is used at most.

**Comment by LucCADORET** [comment]
I would like to know if there are a difference in the use of the connections depending on if you use the `unidici.request()` compared to a client. E.g. if I'm performing 5 requests with `unidici.requests` on the same endpoint, is it more costly that doint the same with a client ?

**Reply by mcollina** [reply]
> E.g. if I'm performing 5 requests with unidici.requests on the same endpoint, is it more costly that doint the same with a client ?

There is as close to no overhead as it's possible. However, your call still passes through a few more code layers, so there will be some.

**Reply by LucCADORET** [reply]
To clarify, what I meant by "overhead" was more like "unoptimized", e.g. in the scenario I described, will the "dispatcher" automagically use the same connection for the 5 requests to this endpoint, or will it create 5 different connections ? Are there any options on the dispatcher to enforce on behavior or the other ? Or is the better option to use the **Client** to have this fine-grained control over connections ?

Overall, I kind of feel a lack of description of the different possible usages of the library, depending on your use case, e.g. "I want a good compromise between throughput of requests and optimized resources usage"

**Reply by mcollina** [reply]
A `Client` wraps to one connection. a `Pool` (or `BalancedPool`) maps to multiple connection to one origin. An `Agent` supports multi-origin via multiple `Pool`s.
Each of them is a dispactcher.

Overall, use an Agent configured with the parameters you need, and you'd be good. There is very little to optimize here minus the parameters in `Agent`.

---

I agree that docs couple have some love. some FAQs would be helpful. Would you like to contribute?

**Reply by LucCADORET** [reply]
Thanks for the info. I'd be happy to contribute once I've fixed my problem (timeouts when running burst of requests in an Azure VM)

**Reply by mcollina** [reply]
Can you describe more what your problems are?

**Reply by LucCADORET** [reply]
On an Azure VM, my program is doing lots of outbound API calls at various intervals, it can go up to 800 at the same instant sometimes
However, when the requests spike arrives, I start to see lots of failed queries (header timeouts, or the queries' signatures get invalidated because the signing requests uses an expired timestamp)
I have tried to limit the connections to 50 using an Agent as a global dispatcher, with pipelining, but it still doesn't handle the spikes properly. I suspect that I run into port exhaustion ( https://learn.microsoft.com/en-us/azure/app-service/troubleshoot-intermittent-outbound-connection-errors )
I wanted to try to use a separate agent **per queried domain** (I query up to 14 different domains) instead of  just one global agent, to fine-tune the amount of connections per domain

**Reply by mcollina** [reply]
I would recommend you to use `BalancedPool` directly. Note that it emits events that allows you to track if there is capacity in the pool: in this way you can sign &send your requests only when there is immediate capacity.

**Reply by LucCADORET** [reply]
Re-implementing with a BalancedPool did not solve our issues, but at this point I am unsure of the exact cause, I need to investigate more. What I did observe though, by looking at the pools `stats` is that the `connected` value stays to 1 even though `connections` in the `BalancedPool` options is set to **128**. Pipelining is set to `10` at the same time, and other values such as `running` shows more than `1` (max observed is 10). So, not sure what `connections` means exactly, if `connected` seems caped to 1
